contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,215 @@
+pub extern crate r2d2;
+
+use std::collections::BTreeMap;
+use std::marker::{Send, Sized};
+use rocket::config::{self, Value};
+
+use self::r2d2::ManageConnection;
+
+#[cfg(feature = "diesel")]
+pub extern crate diesel;
+
+#[cfg(feature = "pg_pool")]
+pub extern crate postgres;
+#[cfg(feature = "pg_pool")]
+pub extern crate r2d2_postgres;
+
+#[cfg(feature = "redis_pool")]
+pub extern crate redis;
+#[cfg(feature = "redis_pool")]
+pub extern crate r2d2_redis;
+
+/// A struct containing database configuration options.
+///
+/// This retrieves the key-value pairs specified in the `Rocket.toml` under
+/// the databases key for use in further configuring the desired connection.
+///
+/// Example configuration:
+/// ```toml
+/// [[global.databases]]
+/// name = "my_database"
+/// url = "postgres://postgres@localhost/my_database
+/// pool_size = 4
+/// ```
+pub struct DatabaseConfig<'a> {
+    /// The connection URL specified in the Rocket configuration
+    pub url: &'a str,
+    /// The size of the pool to be initialized. Defaults to the number of Rocket workers.
+    pub pool_size: u32,
+    /// Any extra options that are included in the configuration.
+    pub extras: Option<&'a BTreeMap<String, Value>>,
+}
+
+/// A custom error type to wrap multiple, disparate error types
+///
+/// This allows for returning the given error from the connection manager
+/// creation methods and the R2D2 connection pool methods.
+#[derive(Debug)]
+pub enum DbError<T> {
+    Custom(T),
+    PoolError(r2d2::Error),
+}
+
+/// This method retrieves the database configuration from the `Rocket.toml` file
+/// and returns a DatabaseConfig struct.
+///
+/// Example usage:
+/// ```rust,ignore
+/// let config = database_config("my_database", rocket.config());
+/// ```
+pub fn database_config<'a>(name: &str, from: &'a config::Config) -> Option<DatabaseConfig<'a>> {
```

Comment:
That actually makes sense to return a `Result` instead of an `Option`. I'll test out your suggestion and correct it as-such.

I agree with keeping it as a `u64`. I think I cast it to a `u32` because I thought it was unreasonable to go above a `u32` number of connections in a pool.